Features/storage constant loss

jnnr opened the pull request:
To implement the functionality of #621, a constant loss term is added to GenericStorage. New attributes `fixed_losses_relative` and `fixed_losses_absolute` are introduced. They are set to zero by default. A scalar or a sequence can be passed.

TODO:
* [x] Docstring formulas,
* [x] an example: https://github.com/oemof/oemof-thermal/blob/features/stratified_thermal_storage/examples/stratified_thermal_storage/example_stratified_thermal_storage.py
* [x] Updated tests.

joroeder commented on oemof/solph/components.py:
```diff
@@ -273,6 +277,8 @@ class GenericStorageBlock(SimpleBlock):
     :math:`c_{max}(t)`          maximum allowed storage :py:obj:`max_storage_level[t]`
     :math:`\delta(t)`           fraction of lost energy :py:obj:`loss_rate[t]`
                                 (e.g. leakage) per time
```
maybe add here `(dependent on E(t)) `

joroeder commented on oemof/solph/components.py:
```diff
@@ -273,6 +277,8 @@ class GenericStorageBlock(SimpleBlock):
     :math:`\delta(t)`           fraction of lost energy :py:obj:`loss_rate[t]`
                                 (e.g. leakage) per time
+    :math:`\gamma(t)`           constant loss of energy :py:obj:`loss_constant[t]`
+                                (e.g. leakage) per time
```
maybe add here `(independent of E(t))`

joroeder commented on tests/test_processing.py:
```diff
@@ -169,6 +169,7 @@ def test_nodes_with_none_exclusion(self):
                 'investment_minimum': 0,
                 'label': 'storage',
                 'inflow_conversion_factor': 1,
+                'loss_constant': 0,
```
I have no idea what test_processing is testing - so probably thats fine, but I don't now.

jnnr replied:
It tests if the return of the outputlib.processing.param_as_dict() function returns the expected values.

p-snft commented:
>  Last dev meeting we were discussing the branch structure and @p-snft made a nice suggestion, I think everybody agreed on... Or is adding a new optional parameter an API change? The old models will still work.

You are right. I rebased the PR to the v0.3 branch.

p-snft commented on oemof/solph/components.py:
```diff
@@ -65,6 +65,9 @@ class GenericStorage(network.Transformer):
         timesteps.
+    loss_constant : numeric (sequence or scalar)
+        A constant loss of the storage level between two consecutive
+        timesteps.
```
As it can be time-dependent, I would call it "fixed loss" instead of "constant loss".

joroeder replied:
I like to call it "fixed losses"

p-snft commented on oemof/solph/components.py:
```diff
@@ -377,6 +384,7 @@ def _storage_balance_rule(block, n, t):
             expr += block.capacity[n, t]
             expr += - block.capacity[n, t-1] * (
                 1 - n.loss_rate[t])
+            expr += n.loss_constant[t]
```
For the loss_rate, relative losses are inherently needed. I think, it is easier here, to have absolute values.

For invest storages, it would even make sense to have to kinds of "fixed" losses: One type growing with the nominal capacity, one that is really fixed.

joroeder replied:
Hm... especially in case of invest storages, I think a capacity dependent fix-loss makes sense. In case of a water tank, whose volume is growing and with the volume the surface, then the fix-losses are growing as well. So, in case if no investment in the storage is done, there will be no fix-losses as well.

jnnr commented:
Thank you for your helpful comments! 

I agree that the losses are better referred to as fixed rather than constant losses. I share @joroeder's view, that the new losses should be in reference to the nominal_storage_capacity. To avoid confusion, I would parallel this in storages and investstorages. So we would have:

* Energy content related loss (`loss_rate`)
* Capacity related losses

As @p-snft suggested, we could have a third type, truly fixed losses. There might be situation where this could be useful. For now, the suggested thermal storages requires capacity related losses (depending on the surface of the storage).

In the meantime, I wonder why the energy content in the storage at each time is called `capacity.` In my view, capacity is what is now called `nominal_storage_capacity`, which tells us the maximum amount of energy that can be stored. What is now called capacity would be better called `energy_content`, no?

p-snft commented:
> In the meantime, I wonder why the energy content in the storage at each time is called capacity. In my view, capacity is what is now called nominal_storage_capacity, which tells us the maximum amount of energy that can be stored. What is now called capacity would be better called energy_content, no?

We have had some discussion about that naming in #538.

jnnr commented:
Thanks for pointing to that. I wrote a short comment there.

pep8speaks commented:
Hello @jnnr! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`oemof/solph/components.py`](https://github.com/oemof/oemof/blob/86120e8902424ecf20cb93be6d736dd509d592bc/oemof/solph/components.py):

> [Line 280:80](https://github.com/oemof/oemof/blob/86120e8902424ecf20cb93be6d736dd509d592bc/oemof/solph/components.py#L280): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 596:80](https://github.com/oemof/oemof/blob/86120e8902424ecf20cb93be6d736dd509d592bc/oemof/solph/components.py#L596): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 615:80](https://github.com/oemof/oemof/blob/86120e8902424ecf20cb93be6d736dd509d592bc/oemof/solph/components.py#L615): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)

jnnr commented:
Thanks again for your helpful comments, @joroeder and @p-snft. I read your feedback and improved the docstrings. The new attribute is now called `fixed_losses`, as you suggested. It is defined as share of the nominal storage capacity. If you find the time, please have another look and tell me if you got other remarks.

jnnr commented:
Hi @FranziPl, I invited you to review as this PR is connected to https://github.com/oemof/oemof-thermal/pull/19

joroeder commented:
sorry, this week is already lot of stuff, hopefully next week I find some time to do a review

FranziPl commented on oemof/solph/components.py:
```diff
@@ -63,8 +63,11 @@ class GenericStorage(network.Transformer):
         timesteps.
+    fixed_losses : numeric (sequence or scalar)
+        Losses as share of nominal storage capacity between two consecutive
+        timesteps independent of storage capacity.
```
Isn't it missleading to call it independent of storage capacity (naming discussion #538). I would say independent of storage_content.

jnnr replied:
>  SoC is well know I think. "storage_content" is misleading to me as well, some could interpret this as water, oil, ...

For discussing this, please go to #538.

jnnr replied:
I changed to 'state of charge' to avoid confusion.

uvchik commented on tests/constraint_tests.py:
```diff
@@ -225,6 +225,7 @@ def test_storage(self):
             outputs={bel: solph.Flow(nominal_value=16667, variable_costs=24)},
             nominal_storage_capacity=10e4,
             loss_rate=0.13,
+            fixed_losses=0.01,
```
Please add a new test instead of changing an existing one.

jnnr replied:
I reverted the commit and will add a new test.

jnnr commented:
> But I have a question: What happens, if the storage is empty, but you still have the fixed losses? How is avoided, that there is a constant input, just to serve this losses?

As you say it. The storage will never be empty as the solver will find a way to avoid that.

jnnr commented:
Thanks for all your comments, very helpful. We noticed that we actually need a third attribute, fixed absolute losses that are not related to nominal storage capacity. I will work on that soon.

jnnr commented:
Thanks again for your helpful comments. I tried to follow your advice and changed the tests as you asked for, @uvchik. In addition, I implemented a further distinction that we discussed before:

> For invest storages, it would even make sense to have to kinds of "fixed" losses: One type growing with the nominal capacity, one that is really fixed.

I did exactly that. Now there are `fixed_losses_relative` and `fixed_losses_absolute`. Some more explanation can be found in the docstring.

Happy to read your feedback!

uvchik commented:
> But I have a question: What happens, if the storage is empty, but you still have the fixed losses? How is avoided, that there is a constant input, just to serve this losses?

The only way to avoid this is a binary variable, which would lead to an MILP but I would open an extra PR if you need it. This will reduce the complexity of each PR.

> I did exactly that. Now there are `fixed_losses_relative` and `fixed_losses_absolute`. Some more explanation can be found in the docstring.

I am interested in a use case for constant absolute fixed losses for an investment storage.
Do you have an example that shows that it works as expected?

I am fine with the tests now :smile:

jnnr commented:
> The only way to avoid this is a binary variable, which would lead to an MILP but I would open an extra PR if you need it. This will reduce the complexity of each PR.

I agree. This can be added if needed, but we prefer not to introduce an integer variable at that point.

> I am interested in a use case for constant absolute fixed losses for an investment storage.
> 
> Do you have an example that shows that your changes work as expected?

There is an example in here: https://github.com/oemof/oemof-thermal/pull/19
If needed, I can add one generic example to the solph-examples in oemof-examples.

joroeder commented:
> > For invest storages, it would even make sense to have to kinds of "fixed" losses: One type growing with the nominal capacity, one that is really fixed.
> 
> I did exactly that. Now there are `fixed_losses_relative` and `fixed_losses_absolute`. Some more explanation can be found in the docstring.

For now, I don't think thats a good idea. In case of an investment storage and the solver is not investing in the storage, there will still be a constant loss. So, there will be a energy loss of a storage without having a storage? E.g. if an invest_relation_input_capacity is given, the solver is forced to build a storage, which is exactly as big as the inflow can cover the losses.
As Uwe said, if we want to consider fix losses, which do not dependent on the nominal_capacity, we need an MILP formulation. Let's skip that for now and think about that in #636 - there we can consider that in a good way.
In case of fix nominal_capacity ("non-invest" case), you can directly calculate the constant losses, I don't understand why additional fixed_losses are necessary in that case - this could be part of the precalculation.

uvchik commented:
> There is an example in here: [oemof/oemof-thermal#19](https://github.com/oemof/oemof-thermal/pull/19)
> If needed, I can add one generic example to the solph-examples in oemof-examples.

Wherever it is convenient for you. 

A link to an example will make the review easier. Otherwise it is time-consuming to check the outcome of the changes. The reviewers have to create their own examples.

jnnr commented:
> For now, I don't think thats a good idea. In case of an investment storage and the solver is not investing in the storage, there will still be a constant loss. So, there will be a energy loss of a storage without having a storage? E.g. if an invest_relation_input_capacity is given, the solver is forced to build a storage, which is exactly as big as the inflow can cover the losses.

Yes, that is a problematic edge case. Good that you point to it! 

> As Uwe said, if we want to consider fix losses, which do not dependent on the nominal_capacity, we need an MILP formulation. Let's skip that for now and think about that in #636 - there we can consider that in a good way.

#636 implements another functionality not directly related to this problem here, but you are right, it is similar in the way that it uses a MILP approach.

> In case of fix nominal_capacity ("non-invest" case), you can directly calculate the constant losses, I don't understand why additional fixed_losses are necessary in that case - this could be part of the precalculation.

Yes, thats right. We could restrict the attribute fixed_losses_absolute to the investment case. But then we had different APIs in the non-invest and invest case, which potentially would cause more confusion.

> One use case could be, that a minimum investment needs to be set as well (in case of investment storage). For example, you want to know how big your storage should be and you already know, that you want to build a storage with a minimum capacity. Then, an additional absolute fixed_losses can make sense. What do you think about that?

This would mean to set another constraint to the already existing one. The existing one is the invest_relation_input_capacity together with the absolute losses. But the new constraint still excludes the option of not building a storage.

uvchik commented:
> Yes, thats right. We could restrict the attribute fixed_losses_absolute to the investment case. But then we had different APIs in the non-invest and invest case, which potentially would cause more confusion.

I would not want different APIs for the invest or non-invest case even though absolute losses and relative losses are the same in the non-invest case.





> There is an example in here: [oemof/oemof-thermal#19](https://github.com/oemof/oemof-thermal/pull/19)
> If needed, I can add one generic example to the solph-examples in oemof-examples.

I get it. Your model of a storage is a tube with a fixed diameter and a variable height. And the losses at the top and the bottom will be the same regardless of the height. In that case you have to define a minimum and maximum invest. Because such an approach needs boarders. In that case the considers of @joroeder are not problematic any more.

jnnr commented:
> I get it. Your model of a storage is a tube with a fixed diameter and a variable height. And the losses at the top and the bottom will be the same regardless of the height. 

Exactly. The losses at the top and bottom are the independent of the invested storage capacity. The `fixed_losses_absolute` could become useful in other storage or battery models as well.

> In that case you have to define a minimum and maximum invest. Because such an approach needs boarders. In that case the considers of @joroeder are not problematic any more.

 Shall I add some check that warns the user when a non-zero `fixed_losses_absolute` is set but no invest_minimum is given?

joroeder commented:
> Shall I add some check that warns the user when a non-zero `fixed_losses_absolute` is set but no invest_minimum is given?

I think thats a nice solution. So people who are using the investment without minimum storage capacity, they are informed that they migth build a constant sink. And, the pre-calculations for investment- and non-investment storage can stay similar. And finally, we already have the ```fixed_absolute_losses``` in the storage API, which we can use as well for develop the milp version where the constant part of the losses is switched off in case of non-investment.

jnnr commented:
Great! I included an error message that checks investment.existing and investment.minimum. 

Thanks again for all your feedback. It there are no more open points, are we ready to merge?
